crl/validator.go

```diff
@@ -20,346 +20,388 @@ package crl
+				if crl != nil || time.Since(crl.lastUpdated) > 4*time.Hour {
+					// Escalate to error logging if the CRL is missing or fails to update in several hours.
+					log.Logger().WithError(err).WithField("CRLDistributionPoint", ep).Error("Update CRL")
+				} else {
+					log.Logger().WithError(err).WithField("CRLDistributionPoint", ep).Debug("Update CRL")
+				}
 			}
-		}
+			wg.Done()
+		}(endpoint, current, wg)
 
-		// If the CRL validator is outdated kill the connection
-		if !v.IsSynced(maxValidityDays) {
-			return errors.New("CRL database is outdated, certificate revocation can't be checked")
-		}
+		return true
+	})
+	wg.Wait()
+}
 
-		return nil
+// updateCRL downloads the CRL from endpoint, and updates the revocationList in crls if it is newer than the current CRL
+func (v *validator) updateCRL(endpoint string, current *revocationList) error {
+	// download CRL
+	crl, err := v.downloadCRL(endpoint)
+	if err != nil {
+		return err
 	}
-}
 
-// parseCRLEndpoints parses the CRLDistributionPoints from registered CA certificates and returns them.
-// Since multiple CA certificates might use the same CRL (although improbable), it is returned as list.
-func (v *validator) parseCRLEndpoints() map[string][]*x509.Certificate {
-	v.certificatesLock.RLock()
-	defer v.certificatesLock.RUnlock()
+	// verify signature
+	if err = v.verifyCRL(crl, current.issuer); err != nil {
+		return err
+	}
 
-	var result = make(map[string][]*x509.Certificate)
-	for _, certificate := range v.certificates {
-		for _, endpoint := range certificate.CRLDistributionPoints {
-			result[endpoint] = append(result[endpoint], certificate)
+	// update when it is a new clr
+	if current.list.Number == nil || current.list.Number.Cmp(crl.Number) < 0 {
+		// parse revocations
+		revoked := make(map[string]bool, len(crl.RevokedCertificates))
+		for _, rev := range crl.RevokedCertificates {
+			revoked[rev.SerialNumber.String()] = true
 		}
+		// set the new CRL
+		v.crls.Store(endpoint, &revocationList{
+			list:        crl,
+			issuer:      current.issuer,
+			revoked:     revoked,
+			lastUpdated: nowFunc(),
+		})
 	}
-
-	return result
+	return nil
 }
 
-// Sync downloads, updates and verifies CRLs
-func (v *validator) Sync() error {
-	endpoints := v.parseCRLEndpoints()
-
-	wc := sync.WaitGroup{}
-	wc.Add(len(endpoints))
-
-	errorsChan := make(chan error)
-
-	go func() {
-		wc.Wait()
-		close(errorsChan)
-	}()
-
-	for endpoint, dependingCAs := range endpoints {
-		if !anyCertificateActive(dependingCAs) {
-			// No active certificates depending on this CRL endpoint
-			wc.Done()
-			continue
-		}
-
-		go func(endpoint string) {
-			defer wc.Done()
-
-			if err := v.updateCRL(endpoint); err != nil {
-				errorsChan <- err
-			}
-		}(endpoint)
+// downloadCRL downloads and parses the CRL
+func (v *validator) downloadCRL(endpoint string) (*x509.RevocationList, error) {
+	response, err := v.httpClient.Get(endpoint)
```

Comment:
what is the time-out?

Reply:
10 seconds? added:
```
// After how long a CRL download attempt is stopped.
// Must be short-ish since the update can happen during a validation request.
const syncTimeout = 10 * time.Second
```